File: typescript/buyer-approval.ts
import { createNexezClient, type AgentPageManifest, type AgentPageOffer } from '@nexez/agent-sdk'

const nexez = createNexezClient({ buyerAgent: 'nexez-buyer-approval-example' })

const proposal = {
  slug: process.env.NEXEZ_APPROVAL_SLUG ?? 'nexez-agent-negotiation-lab',
  offer: process.env.NEXEZ_APPROVAL_OFFER ?? 'services-0',
  query: process.env.NEXEZ_APPROVAL_QUERY ?? 'Buyer wants a one-week agent negotiation sprint.',
  budget: process.env.NEXEZ_APPROVAL_BUDGET ?? 'USD 2100',
  timeline: process.env.NEXEZ_APPROVAL_TIMELINE ?? 'next week',
  requestedTerms: {
    scope: 'Discovery call, agent-readable offer review, and dry-run guidance.',
  },
}
const contactToShare = process.env.NEXEZ_APPROVAL_CONTACT ?? 'buyer@example.com'

const manifest = await nexez.getAgentPage(proposal.slug)
const offer = manifest.offers.find((item) => item.key === proposal.offer)

if (!offer) {
  console.log(`Offer ${proposal.offer} was not found on ${proposal.slug}.`)
  process.exit(0)
}

const dryRun = offer.negotiation_action
  ? await nexez.validateNegotiation(proposal)
  : await nexez.validateCheckout({
      slug: proposal.slug,
      offer: proposal.offer,
      query: proposal.query,
    })

const approval = buildBuyerApprovalSummary({
  actionType: offer.negotiation_action ? 'submit_negotiation' : 'open_checkout',
  manifest,
  offer,
  proposal,
  contactToShare: contactToShare || null,
  dryRun,
})

console.log(JSON.stringify(approval, null, 2))

// In a real buyer agent, render approval.buyer_copy and wait for an explicit
// buyer click/tap/voice confirmation before performing the next action.
const approvedByBuyer = false
if (!approvedByBuyer) {
  console.log('Stopped: buyer approval is required before any side effect.')
  process.exit(0)
}

if (approval.action_type === 'submit_negotiation') {
  const submitted = await nexez.submitNegotiation({
    ...proposal,
    contact: contactToShare,
    approvalToken: getApprovalToken(dryRun),
    userApproved: true,
  }, {
    idempotencyKey: crypto.randomUUID(),
  })
  console.log({
    submitted: {
      ok: submitted.ok,
      status: submitted.status,
      negotiationId: submitted.negotiationId,
      decisionPending: submitted.decisionPending,
    },
  })

  if (submitted.negotiationId && submitted.statusToken) {
    const status = await nexez.waitForNegotiationDecision({
      negotiationId: submitted.negotiationId,
      statusToken: submitted.statusToken,
      timeoutMs: 30_000,
      intervalMs: 2_000,
    })
    console.log({ status })
  }
} else {
  const checkout = await nexez.startCheckout({
    slug: proposal.slug,
    offer: proposal.offer,
    query: proposal.query,
    buyerEmail: contactToShare || undefined,
    approvalToken: getApprovalToken(dryRun),
    userApproved: true,
  }, {
    idempotencyKey: crypto.randomUUID(),
  })
  console.log({
    checkout: {
      provider: checkout.provider,
      url: checkout.url,
    },
  })
}

type ApprovalActionType = 'submit_negotiation' | 'open_checkout'

function getApprovalToken(value: unknown) {
  if (!value || typeof value !== 'object') return undefined
  const token = (value as { approvalToken?: unknown }).approvalToken
  return typeof token === 'string' ? token : undefined
}

type ApprovalSummaryInput = {
  actionType: ApprovalActionType
  manifest: AgentPageManifest
  offer: AgentPageOffer
  proposal: typeof proposal
  contactToShare: string | null
  dryRun: unknown
}

function buildBuyerApprovalSummary(input: ApprovalSummaryInput) {
  const sellerName = input.manifest.page.name
  const offerPrice = input.offer.price ?? 'price not listed'
  const actionLabel =
    input.actionType === 'submit_negotiation' ? 'Approve negotiation submission' : 'Approve checkout handoff'
  const actionDescription =
    input.actionType === 'submit_negotiation'
      ? 'send this proposal to the seller'
      : 'open the seller checkout or booking flow'
  const contactDestination =
    input.actionType === 'submit_negotiation' ? sellerName : input.offer.checkout_url
  const contactNotice = input.contactToShare
    ? ` This will share ${input.contactToShare} with ${contactDestination}.`
    : ''

  return {
    schema_version: 'nexez.buyer-approval.v1',
    requires_buyer_approval: true,
    action_type: input.actionType,
    seller: {
      name: sellerName,
      slug: input.manifest.page.slug,
      public_url: input.manifest.page.url,
      website_url: input.manifest.page.website_url ?? null,
      location: input.manifest.page.location ?? null,
    },
    offer: {
      key: input.offer.key,
      name: input.offer.name,
      price: offerPrice,
      summary: input.offer.voice_summary ?? input.offer.description ?? null,
      checkout_url: input.offer.checkout_url,
    },
    proposal: {
      query: input.proposal.query,
      budget: input.proposal.budget,
      timeline: input.proposal.timeline,
      requested_terms: input.proposal.requestedTerms,
      contact_shared: false,
      contact_share_status: input.contactToShare ? 'pending_approval' : 'not_included',
      contact_to_share: input.contactToShare,
      contact_destination: input.contactToShare ? contactDestination : null,
    },
    dry_run: input.dryRun,
    risk_notes: [
      'No money should move before the buyer approves.',
      'No buyer contact details should be sent before approval.',
      'Dry-run validation may log an analytics attempt, but it does not create checkout, seller contact, or a negotiation.',
    ],
    buyer_copy: {
      title: `${sellerName} - ${input.offer.name}`,
      body: `I found ${input.offer.name} from ${sellerName} at ${offerPrice}. I can ${actionDescription} using your budget (${input.proposal.budget}) and timeline (${input.proposal.timeline}).${contactNotice}`,
      confirmation_question: `Do you approve this ${input.actionType === 'submit_negotiation' ? 'proposal submission' : 'checkout handoff'}?`,
      approve_label: actionLabel,
      cancel_label: 'Cancel',
    },
  }
}

File: python/buyer_approval.py
from __future__ import annotations

import json
import os
from typing import Any, Dict, Optional
from uuid import uuid4

from nexez_agent_sdk import NexezApiError, create_client


def main() -> None:
    nexez = create_client(buyer_agent="nexez-buyer-approval-example")

    proposal = {
        "slug": os.getenv("NEXEZ_APPROVAL_SLUG", "nexez-agent-negotiation-lab"),
        "offer": os.getenv("NEXEZ_APPROVAL_OFFER", "services-0"),
        "query": os.getenv("NEXEZ_APPROVAL_QUERY", "Buyer wants a one-week agent negotiation sprint."),
        "budget": os.getenv("NEXEZ_APPROVAL_BUDGET", "USD 2100"),
        "timeline": os.getenv("NEXEZ_APPROVAL_TIMELINE", "next week"),
        "requested_terms": {
            "scope": "Discovery call, agent-readable offer review, and dry-run guidance.",
        },
    }
    contact_to_share = os.getenv("NEXEZ_APPROVAL_CONTACT", "buyer@example.com")

    manifest = nexez.get_agent_page(proposal["slug"])
    offer = next((item for item in manifest.get("offers", []) if item.get("key") == proposal["offer"]), None)
    if not offer:
        print(f"Offer {proposal['offer']} was not found on {proposal['slug']}.")
        return

    if offer.get("negotiation_action"):
        dry_run = nexez.validate_negotiation(proposal)
        action_type = "submit_negotiation"
    else:
        dry_run = nexez.validate_checkout(
            slug=proposal["slug"],
            offer=proposal["offer"],
            query=proposal["query"],
        )
        action_type = "open_checkout"

    approval = build_buyer_approval_summary(
        action_type=action_type,
        manifest=manifest,
        offer=offer,
        proposal=proposal,
        contact_to_share=contact_to_share or None,
        dry_run=dry_run,
    )

    print(json.dumps(approval, indent=2))

    # In a real buyer agent, render approval["buyer_copy"] and wait for an
    # explicit buyer click/tap/voice confirmation before performing the next action.
    approved_by_buyer = False
    if not approved_by_buyer:
        print("Stopped: buyer approval is required before any side effect.")
        return

    if approval["action_type"] == "submit_negotiation":
        submitted = nexez.submit_negotiation(
            proposal,
            contact=contact_to_share,
            approval_token=dry_run.get("approvalToken"),
            user_approved=True,
            idempotency_key=uuid4().hex,
        )
        print(
            {
                "submitted": {
                    "ok": submitted.get("ok"),
                    "status": submitted.get("status"),
                    "negotiationId": submitted.get("negotiationId"),
                    "decisionPending": submitted.get("decisionPending"),
                }
            }
        )

        negotiation_id = submitted.get("negotiationId")
        status_token = submitted.get("statusToken")
        if negotiation_id and status_token:
            try:
                status = nexez.wait_for_negotiation_decision(
                    negotiation_id,
                    status_token,
                    timeout=30.0,
                    poll_interval=2.0,
                )
                print({"status": status})
            except TimeoutError:
                print({"status": "timed_out", "next": "Check this negotiation again later."})
    else:
        checkout = nexez.start_checkout(
            slug=proposal["slug"],
            offer=proposal["offer"],
            query=proposal["query"],
            buyer_email=contact_to_share or None,
            approval_token=dry_run.get("approvalToken"),
            user_approved=True,
            idempotency_key=uuid4().hex,
        )
        print(
            {
                "checkout": {
                    "provider": checkout.get("provider"),
                    "url": checkout.get("url"),
                }
            }
        )


def build_buyer_approval_summary(
    *,
    action_type: str,
    manifest: Dict[str, Any],
    offer: Dict[str, Any],
    proposal: Dict[str, Any],
    contact_to_share: Optional[str],
    dry_run: Dict[str, Any],
) -> Dict[str, Any]:
    seller_name = manifest["page"]["name"]
    offer_price = offer.get("price") or "price not listed"
    is_negotiation = action_type == "submit_negotiation"
    action_label = "Approve negotiation submission" if is_negotiation else "Approve checkout handoff"
    action_description = "send this proposal to the seller" if is_negotiation else "open the seller checkout or booking flow"
    contact_destination = seller_name if is_negotiation else offer.get("checkout_url")
    contact_notice = (
        f" This will share {contact_to_share} with {contact_destination}."
        if contact_to_share
        else ""
    )

    return {
        "schema_version": "nexez.buyer-approval.v1",
        "requires_buyer_approval": True,
        "action_type": action_type,
        "seller": {
            "name": seller_name,
            "slug": manifest["page"]["slug"],
            "public_url": manifest["page"]["url"],
            "website_url": manifest["page"].get("website_url"),
            "location": manifest["page"].get("location"),
        },
        "offer": {
            "key": offer["key"],
            "name": offer["name"],
            "price": offer_price,
            "summary": offer.get("voice_summary") or offer.get("description"),
            "checkout_url": offer.get("checkout_url"),
        },
        "proposal": {
            "query": proposal["query"],
            "budget": proposal["budget"],
            "timeline": proposal["timeline"],
            "requested_terms": proposal.get("requested_terms"),
            "contact_shared": False,
            "contact_share_status": "pending_approval" if contact_to_share else "not_included",
            "contact_to_share": contact_to_share,
            "contact_destination": contact_destination if contact_to_share else None,
        },
        "dry_run": dry_run,
        "risk_notes": [
            "No money should move before the buyer approves.",
            "No buyer contact details should be sent before approval.",
            "Dry-run validation may log an analytics attempt, but it does not create checkout, seller contact, or a negotiation.",
        ],
        "buyer_copy": {
            "title": f"{seller_name} - {offer['name']}",
            "body": (
                f"I found {offer['name']} from {seller_name} at {offer_price}. "
                f"I can {action_description} using your budget ({proposal['budget']}) "
                f"and timeline ({proposal['timeline']}).{contact_notice}"
            ),
            "confirmation_question": f"Do you approve this {'proposal submission' if is_negotiation else 'checkout handoff'}?",
            "approve_label": action_label,
            "cancel_label": "Cancel",
        },
    }


if __name__ == "__main__":
    try:
        main()
    except NexezApiError as error:
        print({"error": str(error), "status": error.status, "body": error.body})
